src/Microsoft.Azure.SignalR.Common/ServiceConnections/MultiEndpointServiceConnectionContainer.cs

```diff
@@ -115,7 +114,42 @@ public Task WriteAsync(ServiceMessage serviceMessage)
                 throw new AzureSignalRNotConnectedException();
             }
 
-            return Task.WhenAll(routed.Select(s => Connections[s]).Select(s => s.WriteAsync(serviceMessage)));
+            var tasks = new List<Task>();
```

Comment:
Why is this code suddenly so complicated? What exactly is it trying to do?

Reply:
We just want to ensure one of the operations succeeded or all of them failed. And if failed from exceptions, throw them.

Reply:
Why isn't WhenAll enough for that?

Reply:
To reduce latency in some situations. If one endpoint succeed and another endpoint lost the ack message, we have to wait for ack timeout if using WhenAll. The posibility of that will go higher if there're many endpoints.

Reply:
Just use WhenAny in a loop then? Or write a helper to encapsulate that logic